File: bff-services/internal/services/notification_service.go
package services

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"bff-services/internal/api/dto"
	"bff-services/internal/types"
)

type NotificationService interface {
	// Template management
	CreateTemplate(ctx context.Context, payload dto.CreateNotificationTemplateRequest) (*types.HTTPResponse, error)
	GetAllTemplates(ctx context.Context) (*types.HTTPResponse, error)
	GetTemplateById(ctx context.Context, id string) (*types.HTTPResponse, error)
	UpdateTemplate(ctx context.Context, id string, payload dto.UpdateNotificationTemplateRequest) (*types.HTTPResponse, error)
	DeleteTemplate(ctx context.Context, id string) (*types.HTTPResponse, error)

	// User notifications
	CreateUserNotification(ctx context.Context, userID string, payload dto.CreateUserNotificationRequest) (*types.HTTPResponse, error)
	GetUserNotifications(ctx context.Context, userID string, limit, offset int, isRead *bool) (*types.HTTPResponse, error)
	MarkNotificationsAsRead(ctx context.Context, userID string, payload dto.MarkAsReadRequest) (*types.HTTPResponse, error)
	GetUnreadCount(ctx context.Context, userID string) (*types.HTTPResponse, error)
	DeleteUserNotification(ctx context.Context, userID, notificationID string) (*types.HTTPResponse, error)

	// Bulk operations
	SendNotificationToUsers(ctx context.Context, templateID string, payload dto.SendNotificationToUsersRequest) (*types.HTTPResponse, error)
}

type NotificationServiceClient struct {
	baseURL    string
	httpClient *http.Client
}

func NewNotificationServiceClient(baseURL string, httpClient *http.Client) *NotificationServiceClient {
	trimmed := strings.TrimRight(baseURL, "/")
	if httpClient == nil {
		httpClient = &http.Client{Timeout: 10 * time.Second}
	}
	return &NotificationServiceClient{
		baseURL:    trimmed,
		httpClient: httpClient,
	}
}

// Template management methods
func (c *NotificationServiceClient) CreateTemplate(ctx context.Context, payload dto.CreateNotificationTemplateRequest) (*types.HTTPResponse, error) {
	return c.doRequest(ctx, http.MethodPost, "/api/notifications/templates", payload, nil)
}

func (c *NotificationServiceClient) GetAllTemplates(ctx context.Context) (*types.HTTPResponse, error) {
	return c.doRequest(ctx, http.MethodGet, "/api/notifications/templates", nil, nil)
}

func (c *NotificationServiceClient) GetTemplateById(ctx context.Context, id string) (*types.HTTPResponse, error) {
	if id == "" {
		return nil, fmt.Errorf("template id is required")
	}
	path := "/api/notifications/templates/" + url.PathEscape(id)
	return c.doRequest(ctx, http.MethodGet, path, nil, nil)
}

func (c *NotificationServiceClient) UpdateTemplate(ctx context.Context, id string, payload dto.UpdateNotificationTemplateRequest) (*types.HTTPResponse, error) {
	if id == "" {
		return nil, fmt.Errorf("template id is required")
	}
	path := "/api/notifications/templates/" + url.PathEscape(id)
	return c.doRequest(ctx, http.MethodPut, path, payload, nil)
}

func (c *NotificationServiceClient) DeleteTemplate(ctx context.Context, id string) (*types.HTTPResponse, error) {
	if id == "" {
		return nil, fmt.Errorf("template id is required")
	}
	path := "/api/notifications/templates/" + url.PathEscape(id)
	return c.doRequest(ctx, http.MethodDelete, path, nil, nil)
}

// User notification methods
func (c *NotificationServiceClient) CreateUserNotification(ctx context.Context, userID string, payload dto.CreateUserNotificationRequest) (*types.HTTPResponse, error) {
	if userID == "" {
		return nil, fmt.Errorf("user id is required")
	}
	path := "/api/notifications/users/" + url.PathEscape(userID) + "/notifications"
	return c.doRequest(ctx, http.MethodPost, path, payload, nil)
}

func (c *NotificationServiceClient) GetUserNotifications(ctx context.Context, userID string, limit, offset int, isRead *bool) (*types.HTTPResponse, error) {
	if userID == "" {
		return nil, fmt.Errorf("user id is required")
	}

	path := "/api/notifications/users/" + url.PathEscape(userID) + "/notifications"
	query := url.Values{}
	query.Add("limit", fmt.Sprintf("%d", limit))
	query.Add("offset", fmt.Sprintf("%d", offset))
	if isRead != nil {
		query.Add("is_read", fmt.Sprintf("%t", *isRead))
	}
	path += "?" + query.Encode()

	return c.doRequest(ctx, http.MethodGet, path, nil, nil)
}

func (c *NotificationServiceClient) MarkNotificationsAsRead(ctx context.Context, userID string, payload dto.MarkAsReadRequest) (*types.HTTPResponse, error) {
	if userID == "" {
		return nil, fmt.Errorf("user id is required")
	}
	path := "/api/notifications/users/" + url.PathEscape(userID) + "/notifications/read"
	return c.doRequest(ctx, http.MethodPut, path, payload, nil)
}

func (c *NotificationServiceClient) GetUnreadCount(ctx context.Context, userID string) (*types.HTTPResponse, error) {
	if userID == "" {
		return nil, fmt.Errorf("user id is required")
	}
	path := "/api/notifications/users/" + url.PathEscape(userID) + "/notifications/unread-count"
	return c.doRequest(ctx, http.MethodGet, path, nil, nil)
}

func (c *NotificationServiceClient) DeleteUserNotification(ctx context.Context, userID, notificationID string) (*types.HTTPResponse, error) {
	if userID == "" {
		return nil, fmt.Errorf("user id is required")
	}
	if notificationID == "" {
		return nil, fmt.Errorf("notification id is required")
	}
	path := "/api/notifications/users/" + url.PathEscape(userID) + "/notifications/" + url.PathEscape(notificationID)
	return c.doRequest(ctx, http.MethodDelete, path, nil, nil)
}

// Bulk operations
func (c *NotificationServiceClient) SendNotificationToUsers(ctx context.Context, templateID string, payload dto.SendNotificationToUsersRequest) (*types.HTTPResponse, error) {
	if templateID == "" {
		return nil, fmt.Errorf("template id is required")
	}
	path := "/api/notifications/templates/" + url.PathEscape(templateID) + "/send"
	return c.doRequest(ctx, http.MethodPost, path, payload, nil)
}

func (c *NotificationServiceClient) doRequest(ctx context.Context, method, path string, payload interface{}, headers http.Header) (*types.HTTPResponse, error) {
	return doRequest(ctx, c.baseURL, method, path, c.httpClient, payload, headers)
}

File: notification-services/src/services/notificationService.ts
import { v4 as uuidv4 } from 'uuid';
import { NotificationRepository } from '../repositories/notificationRepository';
import { logger } from '../logger';
import {
    NotificationTemplate,
    CreateNotificationTemplate,
    UpdateNotificationTemplate,
    UserNotification,
    CreateUserNotification,
    NotificationTemplateWithCount,
    UserNotificationWithTemplate,
} from '../models/notification';

export class NotificationService {
    private repository: NotificationRepository;

    constructor() {
        this.repository = new NotificationRepository();
    }

    // Notification Template Services
    async createTemplate(templateData: CreateNotificationTemplate): Promise<NotificationTemplate> {
        try {
            const template = await this.repository.createTemplate(templateData);
            logger.info({ templateId: template.id }, 'Notification template created');
            return template;
        } catch (error) {
            logger.error({ error, templateData }, 'Failed to create notification template');
            throw error;
        }
    }

    async getTemplateById(id: string): Promise<NotificationTemplate | null> {
        try {
            return await this.repository.getTemplateById(id);
        } catch (error) {
            logger.error({ error, id }, 'Failed to get notification template');
            throw error;
        }
    }

    async getAllTemplates(): Promise<NotificationTemplateWithCount[]> {
        try {
            return await this.repository.getAllTemplates();
        } catch (error) {
            logger.error({ error }, 'Failed to get all notification templates');
            throw error;
        }
    }

    async updateTemplate(id: string, updates: UpdateNotificationTemplate): Promise<NotificationTemplate | null> {
        try {
            const template = await this.repository.updateTemplate(id, updates);
            if (template) {
                logger.info({ templateId: id }, 'Notification template updated');
            }
            return template;
        } catch (error) {
            logger.error({ error, id, updates }, 'Failed to update notification template');
            throw error;
        }
    }

    async deleteTemplate(id: string): Promise<boolean> {
        try {
            const deleted = await this.repository.deleteTemplate(id);
            if (deleted) {
                logger.info({ templateId: id }, 'Notification template deleted');
            }
            return deleted;
        } catch (error) {
            logger.error({ error, id }, 'Failed to delete notification template');
            throw error;
        }
    }

    // User Notification Services
    async createUserNotification(notificationData: CreateUserNotification): Promise<UserNotification> {
        try {
            // Verify template exists
            const template = await this.repository.getTemplateById(notificationData.notification_id);
            if (!template) {
                throw new Error('Notification template not found');
            }

            const notification = await this.repository.createUserNotification(notificationData);
            logger.info({
                notificationId: notification.id,
                userId: notification.user_id,
                templateId: notification.notification_id
            }, 'User notification created');

            return notification;
        } catch (error) {
            logger.error({ error, notificationData }, 'Failed to create user notification');
            throw error;
        }
    }

    async getUserNotifications(
        userId: string,
        limit: number = 50,
        offset: number = 0,
        isRead?: boolean
    ): Promise<UserNotificationWithTemplate[]> {
        try {
            return await this.repository.getUserNotifications(userId, limit, offset, isRead);
        } catch (error) {
            logger.error({ error, userId, limit, offset, isRead }, 'Failed to get user notifications');
            throw error;
        }
    }

    async markNotificationsAsRead(userId: string, notificationIds: string[]): Promise<number> {
        try {
            const updatedCount = await this.repository.markNotificationsAsRead(userId, notificationIds);
            logger.info({
                userId,
                notificationIds,
                updatedCount
            }, 'Notifications marked as read');
            return updatedCount;
        } catch (error) {
            logger.error({ error, userId, notificationIds }, 'Failed to mark notifications as read');
            throw error;
        }
    }

    async getUnreadCount(userId: string): Promise<number> {
        try {
            return await this.repository.getUnreadCount(userId);
        } catch (error) {
            logger.error({ error, userId }, 'Failed to get unread count');
            throw error;
        }
    }

    async deleteUserNotification(id: string, userId: string): Promise<boolean> {
        try {
            const deleted = await this.repository.deleteUserNotification(id, userId);
            if (deleted) {
                logger.info({ notificationId: id, userId }, 'User notification deleted');
            }
            return deleted;
        } catch (error) {
            logger.error({ error, id, userId }, 'Failed to delete user notification');
            throw error;
        }
    }

    // Bulk operations
    async sendNotificationToUsers(templateId: string, userIds: string[]): Promise<UserNotification[]> {
        try {
            // Verify template exists
            const template = await this.repository.getTemplateById(templateId);
            if (!template) {
                throw new Error('Notification template not found');
            }

            const notifications: UserNotification[] = [];

            for (const userId of userIds) {
                const notification = await this.repository.createUserNotification({
                    user_id: userId,
                    notification_id: templateId,
                });
                notifications.push(notification);
            }

            logger.info({
                templateId,
                userIds: userIds.length,
                notificationsCreated: notifications.length
            }, 'Bulk notifications sent');

            return notifications;
        } catch (error) {
            logger.error({ error, templateId, userIds }, 'Failed to send bulk notifications');
            throw error;
        }
    }
}
